### ISSUE TITLE
Bug: [class-methods-use-this] should be able to ignore implementation of abstract methods

### ISSUE THREAD

**Issue by Zamralik** [typescript-eslint/typescript-eslint#8660]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.4.2&fileType=.tsx&code=IYIwzgLgTsDGEAJYBthjAgguacIDEB7QhAbwFgAoBBABwFcRkBLWBUSGeBEYKACgCUALgQA7egFsQAUygBuKgF8qVFGgxESMgB4QZYgCYZsnPFrJUaDJqx58hoidLmXqNBFBkR6UMQgAWACZFdxVKcKA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6M%2BAQ2WVoFtE%2BABYB7ACa9YyRLWGVUGANqRE0aCOiQANJnDYclAOZMNiAPIA3VVTGIAsoNET0YfHERa9%2ByEZPREAMJcPCgAKkKc%2BACSfMRIAiwAgkxRLKoAZpxkiC5usIheYAC%2BALp6RSBFQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
abstract class AbstractFoo {
  public abstract bar(): number;
}

class Foo extends AbstractFoo {
  public bar(): number {
    return 42;
  }
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/class-methods-use-this": ["error", {
      "ignoreOverrideMethods": true,
      "ignoreClassesThatImplementAnInterface": true
    }]
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I expected the implementation of the method `bar` to not be flagged as an error when `ignoreClassesThatImplementAnInterface` and `ignoreOverrideMethods` are set to `true`.

### Actual Result

The implementation of the method `bar` was flagged as an error because it doesn't have `this` in the body despite `ignoreClassesThatImplementAnInterface` and `ignoreOverrideMethods` being set to `true`.

### Additional Info

Either `ignoreClassesThatImplementAnInterface` or `ignoreOverrideMethods`, when set to `true`, should make this rule skip checking abstract method implementation. `ignoreClassesThatImplementAnInterface` seem more fitting as an abstract class is similar to an interface.

**Comment by auvred** [comment]
From the docs of the `ignoreOverrideMethods` option - https://typescript-eslint.io/rules/class-methods-use-this/#ignoreoverridemethods:

> Makes the rule to ignores any class member explicitly marked with override.

You should add `override` keyword to explicitly state that this method overrides a method from the parent class.

```ts
abstract class AbstractFoo {
  public abstract bar(): number;
}

class Foo extends AbstractFoo {
  public override bar(): number {
    //   ^^^^^^^^
    return 42;
  }
}
```

[playground](https://typescript-eslint.io/play/#ts=5.4.2&fileType=.tsx&code=IYIwzgLgTsDGEAJYBthjAgguacIDEB7QhAbwFgAoBBABwFcRkBLWBUSGeBEYKACgCUALgQA7egFsQAUygBuKgF8qVFGgxESMgB4QZYgCYZsnPFrJUaDJqwSEAbnKjNDMnnyGiJ0uZeo0CAD0QYEAehGREVaBUDIQ9FBiCAAsAEyKASqU2UA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6M%2BAQ2WVoFtE%2BABYB7ACa9YyRLWGVUGANqRE0aCOiQANJnDYclAOZMNiAPIA3VVTGIAsoNET0YfHERa9%2ByEZPREAMJcPCgAKkKc%2BACSfMRIAiwAgkxRLKoAZpxkiC5usIheYAC%2BALp6RSBFQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

**Comment by Zamralik** [comment]
@auvred Except in the case of an abstract method there is no override, it doesn't trigger a TS error when `noImplicitOverride` is enabled. There can only be an override if there is a non-abstract method above in the prototype chain.

**Comment by bradzacher** [comment]
You're thinking in the literal sense of the word "override". However semantically in TS it also means "implementing an abstract method".

Really you should think of the keyword as meaning "there's a thing in the parent that I am creating an implementation for". That "thing" could be a method implementation or it could be an abstract method signature. Either way you are providing an implementation for that thing.

The point of the keyword is to allow TS to error if the base implementation changes and the thing you were "overriding" is removed. If you think of it that way - it makes complete sense that it is valid for an abstract implementation. And it's generally a good practice to add the annotation so you gain the benefits.

That is all an unrelated tangent, however. This rule is not built on top of `noImplicitOverride` - it merely utilises the `override` keyword.

We require for this rule that you must add the keyword to mark a method as existing without `this` on purpose.

**Comment by Zamralik** [comment]
@bradzacher For consistency, that would need a new rule that requires adding `override` when implementing an abstract method.
